### ISSUE TITLE
treesample.py crashes if you press Home key

### ISSUE THREAD

**Issue by pdeblanc** [urwid/urwid#318]
##### Description:
Pressing 'home' or 'end' keys crashes when viewing a TreeWalker widget. This can happen when running the example code in `treesample.py`.

##### Affected versions (if applicable)
- Urwid 2.0.1 running on Python 3.7.0 on macOS Mojave

##### Steps to reproduce (if applicable)
`wget https://raw.githubusercontent.com/urwid/urwid/master/examples/treesample.py`
`python3 treesample.py`
Press `home` key

##### Expected/actual outcome
Expected outcome: nothing happens, since we're already at the top of the tree.

Actual outcome:
```
Traceback (most recent call last):
  File "treesample.py", line 137, in <module>
    main()
  File "treesample.py", line 133, in main
    ExampleTreeBrowser(sample).main()
  File "treesample.py", line 112, in main
    self.loop.run()
  File "/usr/local/lib/python3.7/site-packages/urwid/main_loop.py", line 286, in run
    self._run()
  File "/usr/local/lib/python3.7/site-packages/urwid/main_loop.py", line 384, in _run
    self.event_loop.run()
  File "/usr/local/lib/python3.7/site-packages/urwid/main_loop.py", line 788, in run
    self._loop()
  File "/usr/local/lib/python3.7/site-packages/urwid/main_loop.py", line 825, in _loop
    self._watch_files[fd]()
  File "/usr/local/lib/python3.7/site-packages/urwid/raw_display.py", line 404, in <lambda>
    event_loop, callback, self.get_available_raw_input())
  File "/usr/local/lib/python3.7/site-packages/urwid/raw_display.py", line 502, in parse_input
    callback(processed, processed_codes)
  File "/usr/local/lib/python3.7/site-packages/urwid/main_loop.py", line 411, in _update
    self.process_input(keys)
  File "/usr/local/lib/python3.7/site-packages/urwid/main_loop.py", line 511, in process_input
    k = self._topmost_widget.keypress(self.screen_size, k)
  File "/usr/local/lib/python3.7/site-packages/urwid/container.py", line 1131, in keypress
    return self.body.keypress( (maxcol, remaining), key )
  File "/usr/local/lib/python3.7/site-packages/urwid/treetools.py", line 421, in keypress
    key = self.__super.keypress(size, key)
  File "/usr/local/lib/python3.7/site-packages/urwid/listbox.py", line 1022, in keypress
    return actual_key(self._keypress_max_left())
  File "/usr/local/lib/python3.7/site-packages/urwid/listbox.py", line 1030, in _keypress_max_left
    self.focus_position = next(iter(self.body.positions()))
AttributeError: 'TreeWalker' object has no attribute 'positions'
```

**Comment by jvantuyl** [comment]
It appears that the `positions` function isn't as optional as the documentation claims.  I've worked around this as follows:

```
class FixedTreeWalker(TreeWalker):
    def positions(self, reverse=False):
        """
        implement positions workaround for TreeWalker
        """
        if reverse:
            yield from reversed(list(self.positions()))
            return
        cur = self.focus.get_root().get_widget()
        while cur is not None:
            yield cur.get_node()
            cur = cur.next_inorder()
```

This is not particularly efficient code, but it works for me.

**Comment by tu500** [comment]
This was fixed by #347, right?

**Comment by madebr** [comment]
Spamming the `Home`, `End`, `PgUp`, `PgDn`, `-` and `+` buttons,
this error got thrown:
```
$ python3 treesample.py 
Traceback (most recent call last):
  File "treesample.py", line 137, in <module>
    main()
  File "treesample.py", line 133, in main
    ExampleTreeBrowser(sample).main()
  File "treesample.py", line 112, in main
    self.loop.run()
  File "/home/maarten/.local/lib/python3.7/site-packages/urwid/main_loop.py", line 287, in run
    self._run()
  File "/home/maarten/.local/lib/python3.7/site-packages/urwid/main_loop.py", line 385, in _run
    self.event_loop.run()
  File "/home/maarten/.local/lib/python3.7/site-packages/urwid/main_loop.py", line 790, in run
    self._loop()
  File "/home/maarten/.local/lib/python3.7/site-packages/urwid/main_loop.py", line 827, in _loop
    self._watch_files[fd]()
  File "/home/maarten/.local/lib/python3.7/site-packages/urwid/raw_display.py", line 417, in <lambda>
    event_loop, callback, self.get_available_raw_input())
  File "/home/maarten/.local/lib/python3.7/site-packages/urwid/raw_display.py", line 515, in parse_input
    callback(processed, processed_codes)
  File "/home/maarten/.local/lib/python3.7/site-packages/urwid/main_loop.py", line 412, in _update
    self.process_input(keys)
  File "/home/maarten/.local/lib/python3.7/site-packages/urwid/main_loop.py", line 513, in process_input
    k = self._topmost_widget.keypress(self.screen_size, k)
  File "/home/maarten/.local/lib/python3.7/site-packages/urwid/container.py", line 1131, in keypress
    return self.body.keypress( (maxcol, remaining), key )
  File "/home/maarten/.local/lib/python3.7/site-packages/urwid/treetools.py", line 421, in keypress
    key = self.__super.keypress(size, key)
  File "/home/maarten/.local/lib/python3.7/site-packages/urwid/listbox.py", line 981, in keypress
    return actual_key(self._keypress_max_right((maxcol, maxrow)))
  File "/home/maarten/.local/lib/python3.7/site-packages/urwid/treetools.py", line 470, in _keypress_max_right
    return self.focus_end(size)
  File "/home/maarten/.local/lib/python3.7/site-packages/urwid/treetools.py", line 487, in focus_end
    lastnode = lastwidget.get_node()
AttributeError: 'NoneType' object has no attribute 'get_node'
```
